MNIST Doc (#70)
mlpack/examples: Kaggle/README.md

# MNIST

> ### DESCRIPTION OF THE DATASET

Reviewer: Please be consistent with capitalization of headings.

> MNIST dataset contains images of handwritten digits. It is one of the most common datasets used for image classifcation. It has 60,000 28x28 grayscale images under the training set and 10,000 28x28 grayscale images under the test set. Each pixel has a value between 0 and 255.

Reviewer: ```suggestion
The MNIST dataset contains images of handwritten digits. It is one of the most common datasets used for image classifcation. It has 60,000 28x28 grayscale images under the training set and 10,000 28x28 grayscale images under the test set. Each pixel has a value between 0 and 255.
```

<p  align="center">

<img  alt="Sample Images From The MNIST Database"  src="https://upload.wikimedia.org/wikipedia/commons/2/27/MnistExamples.png">

</p>
<p  align="center">

> <i>Sample Images From The MNIST Database.<i/>

Reviewer: ```suggestion
<i>Sample Images From The MNIST Database.</i>
```

</p>

<br/>

<p  align="center">

> <img  alt="A visualization of the sample image at index 7777"  src="https://miro.medium.com/max/490/1*nlfLUgHUEj5vW7WVJpxY-g.png">

Reviewer: Is this a stable url?

Reviewer: Well, it's been a few months and still looks ok.

</p>
<p  align="center">

> <i>A visualization of the sample image at index 7777.<i/>

Reviewer: ```suggestion
<i>A visualization of the sample image at index 7777.</i>
```

</p>

<br/>

### Processing The Dataset

##### 1. Loading The Dataset From The File To A Useable Format

> The Load Class is used to load a matrix from the file.The type of file is guessed automatically from the extension. For a list of supported file types please refer to [Load Class](https://github.com/mlpack/mlpack/blob/master/src/mlpack/core/data/load.hpp). Please note that since mlpack requires matrices to be in column major format and usually data is stored in a row major format, the matrix by default is transposed. In case the data is originally in column major format, consider setting the parameter to False. Refer to documentation of Load Class for further information.

Reviewer 2: ```suggestion
The Load Class is used to load a matrix from the file. The type of file is guessed automatically from the extension. For a list of supported file types please refer to [Load Class](https://github.com/mlpack/mlpack/blob/master/src/mlpack/core/data/load.hpp). Please note that since mlpack requires matrices to be in column major format and usually data is stored in a row major format, the matrix is transposed by default. In case the data is originally in column major format, consider setting `transpose` to false. Refer to documentation of Load Class for further information.
```

Can we link to the official documentation instead of the file directly?

Author: Sure makes sense.

After loading the dataset is stored in the _tempDataset_ Matrix.

> ```cpp
> mat tempDataset;
> data::Load("Kaggle/data/train.csv", tempDataset, true);

Reviewer 3: ```suggestion
```cpp
mat tempDataset;
data::Load("../data/train.csv", tempDataset, true);
```

```

##### 2. Splitting The Dataset

> The Split Class is used to split the Dataset into Train and Validation Datasets. The train dataset is used in the training procedure, the weights are updated and adjusted to fit this dataset. The validation dataset is used to constantly evaludate the model, which is used to adjust the model hyperparamters like the learning rate. The data is split randomly and RATIO defines the amount of data in Validation Set.

Reviewer 2: ```suggestion
The Split Class is used to split the dataset into training and validation datasets. The training dataset is used in the training procedure. The weights are updated and adjusted to fit this dataset. The validation dataset is used to constantly evaluate the model and adjust the model hyperparameters like the learning rate. The data is split randomly and RATIO defines the amount of data in the validation dataset.
```

> After splitting the train dataset is stored in the _train_ matrix and validation in the _valid_ matrix.

Reviewer 2: ```suggestion
After splitting the training dataset is stored in the _train_ matrix and validation dataset in the _valid_ matrix.
```

```cpp
  mat train, valid;
  data::Split(dataset, train, valid, RATIO);
```

### Results Obtained on Different Models

> ##### 1. DigitRecognizer

Reviewer 3: ```suggestion
##### 1. Mnist
```

> For More Details related to the Model, check out the **DigitRecognizer** Directory.

Reviewer 2: ```suggestion
For more details related to the model, check out the **DigitRecognizer** Directory.
```
Can we link to the directory? Could you also explain what these models actually are?

The same suggestion for everything below

Author: So the idea we discussed (@kartikdutt18) was to explain the dataset and processing along with the results in this README and explain the models in their respective folders, e.g. DigitRecognizerCNN would contain a description of the CNN model.

Reviewer 2: Just a quick overview, so people know what they're looking at. Needn't be in depth at all.

Author: Ok, I will do that.

Author: Is something like this model is a NN with 2 hidden layers and NLL as a loss function or should i add more ?

Reviewer 2: No, I was thinking something more like what Ryan wrote [here](https://github.com/mlpack/examples/blob/92385adcca7cf117875a03cd370dfd0d7e2699a1/README.md)

Reviewer 4: Hi @codeboy5, @favre49, So this is what I think, this repo is also being restructured to remove folders and each model will have its own folder so I believe this is what should be done:
1. Have a notebook like structure that makes more sense for repo like this.

So this would follow this step:

1. A basic overview of dataset (probably an image), shape of image and labels.

2. Steps to implement the model. For example:

>
Step 3: Create the model
We can create a CNN model by stacking Convolutional layer as shown below.
```
   model->Add(layer1);
    ..
   model->Add(layerN);
```

No need to add all layers just the show 1 or 2 cn layers.

> Step 4: Training
One or two lines about using ensmallen callbacks with code.

and so.

Let me know if this makes sense.

Reviewer 4: I'll open documentation for LSTM today you can check that out and maybe that makes sense.

Reviewer 2: Ah, I forgot about this. However, wouldn't it be better in some cases to group some examples together? Like for Digit Recognizers, or for LSTM (basically, like the current file structure). I'm ambivalent on this, but it does fix a problem for when we have, say, 40 or 50 examples. When repos get big, I personally like it when they are well organized

Reviewer 4: That's what I had thought as well. I thought we could have folders like:
```
--mnist_object_classification
  --cnn
  --nn
  --batchnorm
--timeseries_univariate
  --lstm
--timeseries_multivariate
  --lstm
```
This also makes documentation cleaner and comprehensive.
What do you think?
 I think this would also need to be reflected in #69 then.

Reviewer 2: Yeah I was thinking the same too. Though IMO we can have timeseries programs under an LSTM examples folder instead.

Author: Waiting for the documentation of LSTM, will edit this to look similar

Reviewer 4: Hey Saksham, A slight delay will push the docs first thing tomorrow, Finally figured out the bug in another PR so I am wrapping that up. Sorry to keep you waiting.

Author: No problems

> For more details related to the model, check out the **DigitRecognizer** Directory.

Reviewer 3: ```suggestion
For more details related to the model, check out the **mnist** Directory.
```

> After running the model for 50 cycles, the model obtained the best validation set accuracy of **97.2143%**.

Reviewer 2: ```suggestion
After running the model for 50 epochs, the model obtained a validation set accuracy of **97.2143%**.
```

Same suggestions below.

> ##### 2. DigitRecognizerBatchNorm

Reviewer 3: ```suggestion
##### 2. Mnist Batch Norm
```

> For More Details related to the Model, check out the **DigitRecognizerBatchNorm** Directory.

Reviewer 3: ```suggestion
For More Details related to the Model, check out the **mnist_batch_norm** Directory.
```

> After running the model for 50 cycles, the model obatained the best validation set accuracy of **91.5952%**.

Reviewer: I'm probably not up to date on terminology, but I'm not sure what you mean by cycles here? Iterations? Epochs? Some other thing?

Author: Epochs

> ##### 3. DigitRecognizerCNN
> 
> For More Details related to the Model, check out the **DigitRecognizerCNN** Directory.

Reviewer 3: ```suggestion
##### 3. Mnist CNN

For More Details related to the Model, check out the **mnist_cnn** Directory.
```
